File: vc/overlay/viewstart.py
"""View start positions (D-33 item 3): a profile-declared view is set to its start position in the same frame it first
paints, so nothing jumps on camera and sticky elements sit where they will stay.

A profile (profiles/<app>/profile.json) declares rules; each cites its quirk record (D-31):

    "view_start": [{"id": "sticky-lane-headers",             # the quirk record id in specs/quirks/<host>.md
                    "url_regex": "/processes/[0-9a-f-]{36}",  # JS RegExp tested against location.href
                    "selector": ".react-flow__viewport",      # the element that holds the view's position
                    "when": ".react-flow__node-swimlaneLane", # optional: only while this selector exists too
                    "by": "wheel",                            # "wheel" (transform canvas) or "scroll" (scroll box)
                    "x": 0, "y": 0,                           # the start position (translation or scroll offset);
                                                              # omit an axis to leave it alone
                    "wheel_px_per_unit": 0.5,                 # optional first guess, refined in the page
                    "settle_ms": 1500}]                       # optional: re-apply window after the view appears

The take runner installs the rules of the spec's profiles right after the jev session opens, before the warm routes and
the start URL load (loop/vcloop/take_runner.py), in dry runs and takes alike, so a dry run proves the mitigation
(D-40). The page part is viewstart.js; its log (window.__vcViewStart.log) lands in the run's result.json.
"""
import json
import os
import re

_JS_PATH = os.path.join(os.path.dirname(__file__), "viewstart.js")
BY = ("wheel", "scroll")
_KEYS = {"id", "url_regex", "selector", "when", "by", "x", "y", "wheel_px_per_unit", "settle_ms", "why", "quirk"}


class RuleError(ValueError):
    pass


def validate(rule, where="view_start"):
    """Raise RuleError for a rule the page script could not apply; return a clean copy (unknown keys refused)."""
    if not isinstance(rule, dict):
        raise RuleError(f"{where}: a rule must be an object")
    extra = set(rule) - _KEYS
    if extra:
        raise RuleError(f"{where}: unknown keys {sorted(extra)}")
    for k in ("id", "url_regex", "selector"):
        if not isinstance(rule.get(k), str) or not rule[k].strip():
            raise RuleError(f"{where}: `{k}` is required (a non-empty string)")
    try:
        re.compile(rule["url_regex"])
    except re.error as e:
        raise RuleError(f"{where} {rule['id']}: url_regex does not compile: {e}") from None
    if rule.get("by", "scroll") not in BY:
        raise RuleError(f"{where} {rule['id']}: `by` must be one of {BY}")
    if rule.get("x") is None and rule.get("y") is None:
        raise RuleError(f"{where} {rule['id']}: give the start position `x` and/or `y`")
    for k in ("x", "y", "wheel_px_per_unit", "settle_ms"):
        v = rule.get(k)
        if v is not None and (isinstance(v, bool) or not isinstance(v, (int, float))):
            raise RuleError(f"{where} {rule['id']}: `{k}` must be a number")
    if rule.get("wheel_px_per_unit") is not None and rule["wheel_px_per_unit"] <= 0:
        raise RuleError(f"{where} {rule['id']}: wheel_px_per_unit must be > 0")
    out = {k: v for k, v in rule.items() if k in _KEYS - {"why", "quirk"}}
    out.setdefault("by", "scroll")
    return out


def rules_from_profiles(profiles):
    """Every profile's `view_start` rules, validated, in profile order (ids unique across profiles)."""
    out, ids = [], set()
    for p in profiles or []:
        for i, r in enumerate(p.get("view_start") or []):
            r = validate(r, f"profile {p.get('name', '?')} view_start[{i}]")
            if r["id"] in ids:
                raise RuleError(f"view_start id {r['id']!r} is declared twice")
            ids.add(r["id"])
            out.append(r)
    return out


def source(rules):
    """The page script with the rules bound: `(function (rules) {...})([...]);`."""
    with open(_JS_PATH, encoding="utf-8") as f:
        fn = f.read().strip()
    return f"{fn}({json.dumps(list(rules), ensure_ascii=False)});"


def install(call, rules):
    """Install on the filmed tab: `call(method, **params)` is a CDP call on that tab's session.
    Page.enable is required for Page.addScriptToEvaluateOnNewDocument to run in new documents. Returns the script
    identifier, or None when there are no rules."""
    if not rules:
        return None
    src = source(rules)
    call("Page.enable")
    ident = call("Page.addScriptToEvaluateOnNewDocument", source=src).get("identifier")
    call("Runtime.evaluate", expression=src, returnByValue=True)      # the document already loaded
    return ident


LOG_JS = "(window.__vcViewStart && window.__vcViewStart.log.splice(0)) || []"


def read_log(js):
    """Take (and clear) the page's correction log; `js(expr)` evaluates in the filmed tab. [] on any failure."""
    try:
        return list(js(LOG_JS) or [])
    except Exception:  # noqa: BLE001 - a navigating page: the entries of that document are gone anyway
        return []


def summary(entries):
    """Per correction: the rule, from -> to, and whether it landed before the view's first painted frame."""
    out = []
    for e in entries:
        out.append({"id": e.get("id"), "from": e.get("from"), "got": e.get("got"), "rounds": e.get("rounds"),
                    "before_first_paint": e.get("frame") == e.get("seen_frame") == e.get("end_frame", e.get("frame")),
                    "stuck": bool(e.get("stuck"))})
    return out

File: vc/overlay/viewstart.js
/* vc view start: set a view's initial position in the same frame the view first paints (D-33 item 3).
 *
 * Installed by vc/overlay/viewstart.py as a document-start script (Page.addScriptToEvaluateOnNewDocument) and once in
 * the current document. Called with the profile's rules (`view_start` in profiles/<app>/profile.json):
 *
 *   {id, url_regex, selector, when?, by: "wheel"|"scroll", x?, y?, wheel_px_per_unit?, settle_ms?}
 *
 * A MutationObserver sees every element that matches `selector` on a page whose URL matches `url_regex` (and, with
 * `when`, only while that selector also exists). It moves the element to (x, y) at once: its MutationObserver callback
 * runs as a microtask right after the app's DOM commit, before the browser's next rendering step, so the first painted
 * frame of the view already shows the start position. `by`:
 *   - "scroll": a scroll container; scrollLeft/scrollTop are set directly;
 *   - "wheel":  a canvas panned by the page on wheel events (a transform, no scroll container, e.g. a flow chart).
 *               The position is the element's computed transform translation; the script dispatches one synthetic
 *               wheel event per correction (deltaX/deltaY = offset / px-per-unit), reads the result after the app's
 *               microtask render and refines px-per-unit (at most 4 rounds, all before the paint).
 * For `settle_ms` (default 1500) after the element first qualifies, a move by the app itself is undone the same way
 * (re-applied during mount). A wheel or pointer-down that is not ours inside the element (the camera's pan, a user)
 * ends the rule for that element; a new element (a remount, another view) starts again.
 *
 * View-only: nothing is clicked, typed or navigated. Window state: window.__vcViewStart.log, one entry per correction
 * {id, t, frame, seen_frame, from, to, got, rounds}; `frame` counts animation frames, so frame === seen_frame means the
 * position was set before the element's first paint.
 */
(function (rules) {
  'use strict';
  var W = window;
  if (W.__vcViewStart) { W.__vcViewStart.setRules(rules); return; }
  var S = W.__vcViewStart = {rules: [], log: [], frame: 0};
  var armed = typeof WeakMap === 'function' ? new WeakMap() : null;
  var live = [];                 // [el, state] pairs still armed (WeakMap is not iterable)
  var ours = 0;
  var now = function () { return (W.performance && performance.now) ? performance.now() : Date.now(); };
  var microtask = function () { return new Promise(function (r) { Promise.resolve().then(r); }); };

  S.setRules = function (rs) {
    S.rules = (rs || []).map(function (r) {
      var o = {}; for (var k in r) o[k] = r[k];
      try { o.re = new RegExp(r.url_regex || '.*'); } catch (e) { o.re = null; }
      o.settle = r.settle_ms == null ? 1500 : +r.settle_ms;
      o.k = +r.wheel_px_per_unit || 1;
      return o;
    });
    scan();
  };

  // "matrix(a, b, c, d, e, f)" / "matrix3d(..16..)" / "none" -> {x, y}
  S.translation = function (tf) {
    var m = /^matrix(3d)?\(([^)]*)\)/.exec(tf || '');
    if (!m) return {x: 0, y: 0};
    var v = m[2].split(',').map(parseFloat);
    return m[1] ? {x: v[12], y: v[13]} : {x: v[4], y: v[5]};
  };
  var pos = function (el, r) {
    if (r.by === 'scroll') return {x: el.scrollLeft, y: el.scrollTop};
    return S.translation(getComputedStyle(el).transform);
  };
  // how far the element is from the rule's start position, per axis (only the axes the rule names)
  S.offset = function (p, r) {
    return {x: r.x == null ? 0 : p.x - r.x, y: r.y == null ? 0 : p.y - r.y};
  };
  var off = function (d) { return Math.abs(d.x) >= 0.5 || Math.abs(d.y) >= 0.5; };

  var wheel = function (el, dx, dy) {
    var b = el.getBoundingClientRect ? el.getBoundingClientRect() : {left: 0, top: 0, width: 0, height: 0};
    var box = el.parentElement && el.parentElement.getBoundingClientRect ? el.parentElement.getBoundingClientRect() : b;
    ours++;
    try {
      el.dispatchEvent(new WheelEvent('wheel', {deltaX: dx, deltaY: dy, deltaMode: 0, bubbles: true, cancelable: true,
        composed: true, clientX: box.left + box.width / 2, clientY: box.top + box.height / 2}));
    } finally { ours--; }
  };

  var fix = function (el, st) {
    var r = st.rule;
    if (st.busy || st.done || S.frame < (st.hold || 0)) return;
    var p = pos(el, r), d = S.offset(p, r);
    if (!off(d)) return;
    st.busy = true;
    var entry = {id: r.id, t: Math.round(now()), frame: S.frame, seen_frame: st.seen, from: p, to: {x: r.x, y: r.y},
                 got: p, rounds: 0};
    var finish = function () { entry.got = pos(el, r); entry.end_frame = S.frame; S.log.push(entry); st.busy = false; };
    if (r.by === 'scroll') {
      if (r.x != null) el.scrollLeft = r.x;
      if (r.y != null) el.scrollTop = r.y;
      entry.rounds = 1; finish(); return;
    }
    var round = function () {
      if (entry.rounds >= 4) return finish();
      var before = pos(el, r), dd = S.offset(before, r);
      if (!off(dd)) return finish();
      entry.rounds++;
      // the page moves the canvas against the wheel: +delta moves the translation by -k*delta
      wheel(el, dd.x / r.k, dd.y / r.k);
      microtask().then(microtask).then(function () {
        var after = pos(el, r), mx = before.x - after.x, my = before.y - after.y;
        var ax = Math.abs(dd.x) >= 0.5 ? mx / (dd.x / r.k) : null, ay = Math.abs(dd.y) >= 0.5 ? my / (dd.y / r.k) : null;
        var k = ax && ax > 0 ? ax : (ay && ay > 0 ? ay : null);
        if (!k) { entry.stuck = true; st.hold = S.frame + 2; return finish(); }   // not moved (yet): retry 2 frames on
        r.k = k;
        round();
      });
    };
    round();
  };

  var scan = function () {
    if (!S.rules.length || !W.document || !document.documentElement) return;
    var href = String(location.href), t = now();
    for (var i = 0; i < S.rules.length; i++) {
      var r = S.rules[i];
      if (!r.re || !r.re.test(href) || !r.selector) continue;
      if (r.when && !document.querySelector(r.when)) continue;
      var els = document.querySelectorAll(r.selector);
      for (var j = 0; j < els.length; j++) {
        var el = els[j], st = armed ? armed.get(el) : el.__vcViewStartState;
        if (!st) {
          st = {rule: r, t0: t, seen: S.frame, done: false};
          if (armed) armed.set(el, st); else el.__vcViewStartState = st;
          live.push([el, st]);
        }
        if (st.done) continue;
        if (t - st.t0 > r.settle) { st.done = true; continue; }
        fix(el, st);
      }
    }
    live = live.filter(function (p) { return !p[1].done && p[0].isConnected !== false; });
  };
  S.scan = scan;

  // input that is not ours inside an armed element (the camera's pan, a person) ends that element's rule
  var foreign = function (e) {
    if (ours) return;
    for (var i = 0; i < live.length; i++) {
      var el = live[i][0];
      if (e.target && (e.target === el || (el.contains && el.contains(e.target)))) live[i][1].done = true;
    }
  };
  W.addEventListener('wheel', foreign, {capture: true, passive: true});
  W.addEventListener('pointerdown', foreign, {capture: true, passive: true});

  var tick = function () {
    S.frame++;
    if (live.length) scan();           // a scroll container can move without a DOM mutation
    requestAnimationFrame(tick);
  };
  requestAnimationFrame(tick);

  var start = function () {
    new MutationObserver(scan).observe(document.documentElement, {subtree: true, childList: true, attributes: true,
                                                                   attributeFilter: ['style', 'class']});
    scan();
  };
  if (document.documentElement) start();
  else document.addEventListener('readystatechange', start, {once: true});
  S.setRules(rules);
})
